crates/asm/txs/bridge-v1/src/unstake/script.rs

```diff
@@ -0,0 +1,299 @@
+        // Verify size of preimage is 32 bytes
+        .push_opcode(OP_SIZE)
+        .push_int(0x20)
+        .push_opcode(OP_EQUALVERIFY)
+        // Verify the preimage matches the hash
+        .push_opcode(OP_SHA256)
+        .push_slice(stake_hash)
+        .push_opcode(OP_EQUAL)
+        .into_script()
+}
+
+/// Extracts the two dynamic parameters (pubkey and stake_hash) from a stake connector script.
+///
+/// This is a minimal extraction that only validates the 32-byte push instructions exist
+/// at the expected positions (0 and 6). Full structural validation happens by reconstructing
+/// the script and comparing byte-for-byte.
+///
+/// Returns `None` if the basic structure doesn't allow parameter extraction.
+fn extract_script_params(script: &ScriptBuf) -> Option<([u8; 32], [u8; 32])> {
+    let mut instructions = script.instructions();
+
+    // Extract pubkey from instruction 0 (first push)
+    let pubkey = match instructions.next() {
+        Some(Ok(Instruction::PushBytes(bytes))) if bytes.len() == 32 => {
+            bytes.as_bytes().try_into().ok()?
+        }
+        _ => return None,
+    };
+
+    // Skip instructions 1-5 (static opcodes - will be validated via reconstruction)
+    for _ in 0..5 {
+        instructions.next();
+    }
+
+    // Extract stake_hash from instruction 6
+    let stake_hash = match instructions.next() {
+        Some(Ok(Instruction::PushBytes(bytes))) if bytes.len() == 32 => {
+            bytes.as_bytes().try_into().ok()?
+        }
+        _ => return None,
+    };
+
+    Some((pubkey, stake_hash))
+}
+
+/// Validates a stake connector script and extracts its parameters.
+///
+/// This function performs complete validation by:
+/// 1. Extracting the pubkey and stake_hash from the script
+/// 2. Reconstructing what the script SHOULD be with those parameters
+/// 3. Comparing byte-for-byte with the original script
+///
+/// Returns the extracted parameters only if the script exactly matches the canonical
+/// `stake_connector_script` output. This ensures the script structure is correct.
+///
+/// # Returns
+/// - `Some((pubkey, stake_hash))` if the script is valid and matches the canonical structure
+/// - `None` if the script is malformed or doesn't match the expected structure
+pub fn validate_and_extract_script_params(
+    script: &ScriptBuf,
+) -> Option<(XOnlyPublicKey, [u8; 32])> {
+    // STEP 1: Extract the two dynamic parameters
+    let (pubkey_bytes, stake_hash_bytes) = extract_script_params(script)?;
+
+    // STEP 2: Parse pubkey to ensure it's a valid X-only public key
+    let pubkey = XOnlyPublicKey::from_slice(&pubkey_bytes).ok()?;
+
+    // STEP 3: Reconstruct what the script SHOULD be
+    let expected_script = stake_connector_script(stake_hash_bytes, pubkey);
+
+    // STEP 4: Byte-for-byte comparison - only return params if script matches exactly
+    if script == &expected_script {
+        Some((pubkey, stake_hash_bytes))
+    } else {
+        None
+    }
```

Comment:
You should be able to do this instead.

```suggestion
    (script == &expected_script).then(|| (pubkey, stake_hash_bytes))
```